src/rdf4cpp/rdf/Literal.cpp

```diff
@@ -157,78 +161,116 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
+                                                                            other_datatype);
 
-    auto const &[res_identifier, lhs_val, rhs_val] = *after_conversion;
+        if (!equalizer.has_value()) {
+            return Literal{}; // not convertible
+        }
 
-    auto const result_entry = datatypes::registry::DatatypeRegistry::get_entry(res_identifier).value().get();
+        auto const equalized_entry = [&]() {
+            if (equalizer->target_type_iri == this_datatype) {
+                return this_entry;
+            }  else if (equalizer->target_type_iri == other_datatype) {
+                return other_entry;
+            } else {
+                return DatatypeRegistry::get_entry(equalizer->target_type_iri);
+            }
+        }();
 
-    if (!result_entry.numeric_ops.has_value()) {
-        return Literal{}; // common datatype is not numeric
-    }
+        assert(equalized_entry.has_value());
+        assert(equalized_entry->get().numeric_ops.has_value());
 
-    auto op_res = op_select(*result_entry.numeric_ops)(lhs_val, rhs_val);
+        DatatypeRegistry::NumericOpResult const op_res = op_select(*equalized_entry->get().numeric_ops)(equalizer->convert_lhs(this->value()),
+                                                                                                        equalizer->convert_rhs(other.value()));
 
-    return Literal {
-        result_entry.to_string_fptr(op_res),
-        IRI{ result_entry.datatype_iri, node_storage },
-        node_storage
-    };
+        auto const to_string_fptr = [&]() {
+            if (op_res.result_type_iri == equalized_entry->get().datatype_iri) [[likely]] {
+                return equalized_entry->get().to_string_fptr;
+            } else [[unlikely]] {
+                return DatatypeRegistry::get_to_string(op_res.result_type_iri);
+            }
+        }();
+
+        assert(to_string_fptr != nullptr);
+
+        return Literal{to_string_fptr(op_res.result_value),
+                       IRI{ op_res.result_type_iri, node_storage },
+                       node_storage};
+    }
 }
 
 template<typename OpSelect>
 Literal Literal::numeric_unop_impl(OpSelect op_select, NodeStorage &node_storage) const {
+    using datatypes::registry::DatatypeRegistry;
+
     if (this->null()) {
         return Literal{};
     }
 
-    assert(this->handle_.is_literal());
-
-    auto const &entry = datatypes::registry::DatatypeRegistry::get_entry(this->datatype().identifier()).value().get();
+    auto const entry = DatatypeRegistry::get_entry(this->datatype().identifier());
+    assert(entry.has_value());
 
-    if (!entry.numeric_ops.has_value()) {
+    if (!entry->get().numeric_ops.has_value()) {
         return Literal{}; // datatype not numeric
     }
 
-    auto op_res = op_select(*entry.numeric_ops)(this->value());
+    DatatypeRegistry::NumericOpResult const op_res = op_select(*entry->get().numeric_ops)(this->value());
```

Comment:
`std::optional<std::reference_wrapper<X>>` is very verbose. Why not a raw pointer? Especially considering that the optional is not checked anyway.

Reply:
Yes, you are right. I was thinking in rust not c++ there 😅 (in rust Option<&T> is the idiomatic thing)